### DISCUSSION TITLE
How do I support Product.Generated in WiX 4 and 5?

### DISCUSSION THREAD

**Discussion by NiceLuc** [wixtoolset/issues#8845] [Questions]
I'm working to upgrade 195 wixproj files from v3 to v5. I go through the process of using the HeatWave extension for Visual Studio 2022.
> [HeatWave for VS2022 - Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=FireGiant.FireGiantHeatWaveDev17)

When clicking on the the **Upgrade to WiX v4...** menu item, it shows success.
![image](https://github.com/user-attachments/assets/feceb9a4-db53-485e-8da9-333fe0c9a9b9)

However, when running the installer, I was getting an error stating the `ComponentGroup:Product.Generated` is not a valid token. As a result, I removed the element and the build passed. However, now the installer does not contain any of the `<ProjectReference />` artifacts. 

I've tried many things, and nothing is working. Wondering if anyone can help me with this? 

After looking through my obj/bin directory, I see this file: `Product.Generated.wxs`, but the element has no child nodes:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <Fragment>
    <ComponentGroup Id="Product.Generated" />
  </Fragment>
</Wix>
```

What is the proper way to fix this using the WixToolset? Thanks for any help!!

**Comment by wmanning** [comment]
In your original 3.x wxs files do you have a `<componentgroup>` element with Id set to "Product.Generated"?
Also, IIRC, it can say success but they'll be a file or something generated that will list warnings. Should take a look at that.

**Reply by NiceLuc** [reply]
@wmanning - Thank you for your response. Yes, we do have a `<ComponentGroupRef Id="Product.Generated" />` in our original (WiX 3) `Product.wxs` files. This has caused big problems with our migration to WiX 4. I'm currently trying to figure out how to use the `<ProjectReference />` elements to get the assets from the various build processes. This is my current attempt, but it's not working for me:

## Installer.wixproj

```xml
  <PropertyGroup>
    <EnableProjectHarvesting>True</EnableProjectHarvesting>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\..\..\C#\Library\Common\Common.csproj">
      <Name>Common</Name>
      <Project>{07c88c80-7a22-413c-a683-9a6b6a75ab84}</Project>
      <Private>True</Private>
      <RefProjectOutputGroups>Binaries;Symbols</RefProjectOutputGroups>
      <RefTargetDir>Binaries</RefTargetDir>
      <ReferenceOutputAssembly>False</ReferenceOutputAssembly>
    </ProjectReference>

   <!-- many others... -->

</ItemGroup>
```

.. I am trying to get the output of these project references like so:

## Product.wxs

```xml
<Wix ...>
   <Package ...>

       <!-- omitted for brevity -->

        <Feature Id="ProductFeature" ...>
            <!-- omitted for brevity -->

            <!-- new way of referencing binaries from ProjectReference elements -->
            <ComponentGroupRef Id="Common.Binaries" />
            <ComponentGroupRef Id="Common.Symbols" />
        </Feature>

        <Directory Id="Binaries" Name="bin"></Directory>

   </Package>
</Wix>   
```

This is not working. 

What is the proper way to get the files generated for Binaries and Symbols from the `<ProjectReference />` elements? Thanks for any additional guidance on this.

**Comment by wmanning** [comment]
I see you're doing harvesting.  Sorry can't help there as I manually maintain mine.

Does all your wix component elements have a File attribute?  That's required in Wix 5 but not in Wix 3.  See here: https://github.com/wixtoolset/issues/issues/8148

**Comment by NiceLuc** [comment]
# Still Blocked

No worries, @wmanning! Thanks for your responses. Is anyone else able to help with this?

**Reply by robmen** [reply]
Note: FireGiant not only provides the HeatWave community edition you are using for free but also [offers commercial support](https://www.firegiant.com/services/support-programs/) that gets you into a support system that will always get you unblocked rapidly. Everyone in it loves the response time and depth of answers for all things related to software installation on Windows, including all versions of the WiX toolset.

**Reply by NiceLuc** [reply]
Thanks @robmen.  I've created a bug report about the issues I'm having. 
> https://github.com/wixtoolset/issues/issues/8858

The logging did not mention that I had "extra work to do" when using the `EnableProjectHarvesting` element. As a result, I learned a _LOT_ over the last few weeks while attempting to fix this. @robmen's video series helped a lot!! 
> https://robmensching.com/deployment-dojo/episodes/s1/

Ultimately, my issue came down to the fact that adding the new `<ComponentGroupRef />` elements _without_ adding the **Heat** gave me a lot of errors and led me in the wrong direction. Once I added the **Heat** package, and massaged my markup a bit, it worked.

Hoping that taking my bug into consideration will help the next person facing these problems.
